File: cpp/src/advanced/basic_namespace.cpp
#include "basic_namespace.h"
using namespace std;
// definizione del namespace
namespace math_operations {
  double somma(double a, double b) { return a + b; }
  double sottrazione(double a, double b) { return a - b;  }
  double moltiplicazione(double a, double b) { return a * b; }
  double divisione(double a, double b) { return (b==0) ? 0 : a/b ; }
}

void namespace_example() {
  double x = 10.5, y = 2.5;
  cout << "Somma: " << math_operations::somma(x, y) << endl;
  cout << "Sottrazione: " << math_operations::sottrazione(x, y) << endl;
  cout << "Moltiplicazione: " << math_operations::moltiplicazione(x, y) << endl;
  cout << "Divisione: " << math_operations::divisione(x, y) << endl;
}

File: rust/src/advanced/basic_namespace.rs

// in rust tutto è privarto per impostazione predefinita,
// per rendere visdibile all'esterno serve la keywork pub

mod math_operations {
  pub mod algebra {
    pub fn somma(a: f32, b: f32) -> f32 { a + b }
    pub fn sottrazione(a: f32, b: f32) -> f32 { a - b }
  } // both syntax permitted and nested namespeces
  pub fn moltiplicazione(a: f32, b: f32) -> f32 { a * b }
  pub fn divisione(a: f32, b: f32) -> f32 { if b==0.0 {0.0} else {a/b}  }
}




pub fn namespace_test() {
  let x = 10.5; let y = 2.5;
  
  println!("Somma: {}", math_operations::algebra::somma(x, y));
  println!("Sottrazione: {}", math_operations::algebra::sottrazione(x, y));
  println!("Moltiplicazione: {}", math_operations::moltiplicazione(x, y));
  println!("Divisione: {}", math_operations::divisione(x, y));
}